### ISSUE TITLE
E0460 long description is too optimistic about cargo's abilities

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#110120]
### Location

https://github.com/rust-lang/rust/blob/5ecac8ede670ac8045a199cd3bcd58bd940ec8d4/compiler/rustc_error_codes/src/error_codes/E0460.md#L61-L67

### Summary

This makes the claim "using cargo will automatically fix your problem". This is unfortunately not true. Consider the following error found while working on https://github.com/rust-lang/rust/issues/81930:

```
error[E0460]: found possibly newer version of crate `core` which `itoa` depends on
   --> /home/jyn/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/serde_json-1.0.85/src/number.rs:295:49
    |
295 |             N::PosInt(u) => formatter.write_str(itoa::Buffer::new().format(u)),
    |                                                 ^^^^
    |
    = note: perhaps that crate needs to be recompiled?
    = note: the following crate versions were found:
            crate `core`: /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-5670a3f7cd4c9169.rmeta
            crate `itoa`: /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/libitoa-c25ddfbc66feadac.rmeta
```

All the dependencies involved here were compiled by cargo, but the error still appears. I am currently trying to figure out why; I think it's related to rebuilding the sysroot, but cargo not noticing that it needs to recompile a crate that depends on `core`? Maybe rustc needs to add a dependency edge to the `.d` files it emits?

cc https://github.com/rust-lang/rust/issues/76720#issuecomment-1369507582, @Ezrashaw

**Comment by jyn514** [comment]
BTW the ASCII art in the long error was extremely helpful, thank you :heart: that's how I figured out it might be related to rebuilding the sysrootl.

**Comment by Ezrashaw** [comment]
Gah, I think I copied that statement from some other docs I wrote without really thinking about it. I do think that, generally speaking, `cargo` should alleviate these woes, just not in all cases it seems. Unfortunately, I'm not sure I'll be able to help that much with this (I just wrote the docs :P) but I'll have a go at getting this error with cargo outside of the compiler.

**Comment by jyn514** [comment]
I think it's going to be hard to replicate, you'll need to find some bug in the integration between rustc and cargo. "not rebuilding when crates in the sysroot are updated" is the only one I know of today, and that only happens with a third build orchestration tool touching the sysroot.

**Comment by Ezrashaw** [comment]
Yeah, so maybe the statement is true if you are just using cargo without bootstrap getting in the way?
